Don't rely on FDT for boot properties, clean up FDT prior to running unix

r1mikey opened the pull request:
To make the startup code reusable between devicetree and ACPI we need to reduce our dependency on FDT for communicating boot properties between inteboot (or any loader shim) and the kernel boot code.

Use a combination of xboot_info and boot modules to disentangle this part of the codebase.

Since the illumos FDT integration requires a phandle for all nodes, and libfdt only insists on phandle for referenced nodes, have the `inetboot` shims tweak the FDT before passing it through to illumos.

Also have the `inetboot` shims ensure that well known pseudo-nodes exist before starting `unix`.

While I'm here, make all files touched pbchk clean.

This diff introduces a few new XXXARM blocks, which are in place to track places we need to adjust to get parity with i86pc.

Tested on both Qemu and Pi4 - Pi4 boot log and prtconf in https://gist.github.com/r1mikey/82e558bb3da5e57cdd492648e7bbcd58 (lightly edited to clean up some bootrd logspew so that it's small enough for a gist).

r1mikey committed:
Don't rely on FDT for boot properties

To make the startup code reusable between devicetree and ACPI we
need to reduce our dependency on FDT for communicating boot properties
between inteboot (or any loader shim) and the kernel boot code.

Use a combination of xboot_info and boot modules to disentangle this
part of the codebase.

Since the illumos FDT integration requires a phandle for all nodes,
and libfdt only insists on phandle for referenced nodes, have the
`inetboot` shims tweak the FDT before passing it through to illumos.

Also have the `inetboot` shims ensure that well known pseudo-nodes exist
before starting `unix`.

While I'm here, make all files touched pbchk clean.

This diff introduces a few new XXXARM blocks, which are in place to
track places we need to adjust to get parity with i86pc.

richlowe commented on usr/src/uts/armv8/os/ddi_impl.c:
```diff
@@ -989,40 +1303,64 @@ impl_ddi_sunbus_removechild(dev_info_t *dip)
 
 		length = BOP_GETPROPLEN(bootops, property_name);
-		if (length == 0)
+		if (length < 0)
```
What does `length == 0` end up doing?

richlowe replied:
Boolean properties, I meant this as a note to myself to double check and then forgot to delete it.  Sorry.

richlowe commented on usr/src/uts/armv8/os/ddi_impl.c:
```diff
@@ -1031,31 +1369,84 @@ get_boot_properties(void)
+		 * bios-boot-device and stdout
+		 *	goes to hardware property list so it may show up
+		 *	in the prtconf -vp output. This is needed by
+		 *	Install/Upgrade. Once we fix install upgrade,
```
Is this comment from i86pc?  We should probably find a better name than `bios-boot-device` if it's relevant to us, but I'm not sure how it could be right now.

r1mikey replied:
Yeah, the code is almost completely transposed from i86pc - with a handful of Arm adjustments.

I need to flesh out my `loader64.efi` work before I can tell if this would be relevant to us, but at first blush I don't think it would be.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -239,22 +293,49 @@ void bop_init(struct xboot_info *xbp)
+	DBG_MSG("Initializing boot properties:\n");
+	build_boot_properties(xbp);
+
+#if XXXARM
```
Is there a reason we can't do this now?

r1mikey replied:
This is the start of a stream of work to sort out how we do the bsvc UART and introduce the text console, which will be backed by either simplefb or the UEFI framebuffer as appropriate for the platform.

With that background, `find_boot_prop` lives in `boot_console.c`, so when I reintroduce that I'll enable these blocks.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+	/*
+	 * align the size to 16 byte boundary
+	 */
+	size = sizeof (bootprop_t) + nlen + 1 + vlen;
```
You should be able to do this with `P2ALIGN`  from `sys/sysmacros.h` I think?

r1mikey replied:
Will do - this was transposition from i86pc.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+
+	/*
+	 * align new values of curr_page, curr_space
+	 */
```
`P2ROUNDUP` and `P2ALIGN` I think?

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+/*
+ * to find the size of the buffer to allocate
+ */
+int
```
should this be `size_t`?

r1mikey replied:
I can update it - it's transposed from i86pc at the moment.

r1mikey replied:
Going to leave this as-is for now. i86pc has it as int as well, and almost all of the callsites expect a signed return value (for error checking).

Perhaps `ssize_t`?

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+}
+
+/*
+ * to find the size of the buffer to allocate
```
is this comment copy/paste error from below, or true of both?

r1mikey replied:
Copy/pasta - I'll update it.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+/*
+ * Parse numeric value from a string. Understands decimal, hex, octal, - and ~
+ */
+static int
```
Could we use `ddi_strtoul` and then just handle the radix and `~` ?

r1mikey replied:
Yeah, that seems reasonable. Again, what's here is just a transposition of what's in i86pc.

I do think there's a ton that can be shared between the platforms (from this sort of code all the way to most of the PCI code).

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -454,6 +876,15 @@ read_bootenvrc(void)
 	}
+	/* XXXARM: early_allocation = 0; */
+
+#if XXXARM
```
Is the comment saying why this can't be live now?  It seems reasonable to have it

r1mikey replied:
This is disabled while we get the boot console up. The commented code above is something I need to dig into in early i86pc memory management to see if we need it here.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -479,107 +909,55 @@ read_bootenvrc(void)
-	pnode_t chosen = prom_chosennode();
-	return prom_getproplen(chosen, name);
-}
+#if XXXARM
```
I assume this is a bit of console stuff from x86 land we can't do yet

r1mikey replied:
Yes, this is related to the boot console - I'm working towards bringing the text console up.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -479,107 +909,55 @@ read_bootenvrc(void)
-	pnode_t chosen = prom_chosennode();
-	prom_getprop(chosen, name, (caddr_t)value);
-	return 0;
+#if XXXARM
```
I don't know why we can't do this part yet?

r1mikey replied:
See the previous comment - I'm trying to not diverge too far from the i86pc structure, so when we get `boot_console.c` I'll enable this.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -479,107 +909,55 @@ read_bootenvrc(void)
-vbop_printf(void *ptr, const char *fmt, va_list ap)
+vbop_printf(void *ptr __unused, const char *fmt, va_list ap)
 {
+#if XXXARM
```
Exciting, I didn't realize we could get here without _any_ console, even on x86.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -479,107 +909,55 @@ read_bootenvrc(void)
  * Another panic() variant; this one can be used even earlier during boot than
  * prom_panic().
  */
-/*PRINTFLIKE1*/
```
You could add the gcc attribute with `__PRINTFLIKE(1)` in the declaration, so it still error checks.

r1mikey replied:
Yeah, another transposition from i86pc - I'll clean it up.

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -592,54 +970,485 @@ bop_panic(const char *fmt, ...)
+		if (strcmp(pn, "phandle") == 0 ||
+		    strcmp(pn, "impl-arch-name") == 0 ||
+		    strcmp(pn, "mfg-name") == 0 ||
+		    strcmp(pn, "__bootpath") == 0) {
```
I'm not sure `__bootpath`  actually gets used for anything, it could just be removed I think

r1mikey replied:
Yeah, it seems unused. The loaders are oddly insistent on setting it :-)

richlowe commented on usr/src/uts/armv8/os/fakebop.c:
```diff
@@ -592,54 +970,485 @@ bop_panic(const char *fmt, ...)
+ * well-handled as bootparams, such as spaces, brackets, and the like, so that's
+ * another reason.
+ */
+static struct bop_blocklist {
```
I guess this is from x86, and assumes what loader would do if we had it?

r1mikey replied:
Yes, from i86pc. I have a port of loader64.efi, which sets a lot of this stuff.

For that port, I'm working to straighten out how we can recognise and reason about ttya/ttyb etc. (since we don't have a "standard I/O port addressing" scheme to fall back on).

richlowe reviewed (commented):
To the degree some of this is now common with x86, we should think in the future about how to actually share the code.  It's probably best to wait until we have a real boot loader so we actually know for sure what we can share though.

r1mikey commented:
Thanks for the review.

To put this all in context, I'm working on cleaning up the contact between the loader shim and `unix` while reducing dependence on FDT implementation details.

A part of this is with an eye on getting ACPI/UEFI support without copying everything we have for FDT. Another goal is to get the framebuffer up. Eventually we'll get PCIe up, at which point console=text should be minimally usable on Pi4.

richlowe commented:
It's looking good, and it should give us a starting point to work out which code can become common or be refactored into being more common.  Thanks for doing this stuff!

r1mikey commented:
I've pulled in the bits of boot_console that are needed for reading the boot properties in fakebop and addressed review comments. Retested on qemu, both with and without prom_debug and kbm_debug (use `-append "-D /virtio_mmio@a003c00 -Bkbm_debug=,prom_debug="`).

r1mikey commented:
On `__bootpath` - this seems no longer relevant.

The `__bootpath` property is used to create `prom_bootpath` in the aarch64 loaders. It's not directly used in the kernel itself.

There are four references to `prom_bootpath` in the tree:
* usr/src/stand/lib/sock/socket.c (only used when `net-config-strategy` is too big(!) to fit into `net_options`)
* usr/src/uts/common/io/strplumb.c (only in the _OBP case, so irrelevant)
* usr/src/uts/sparc/os/cpr_sparc.c (SPARC, so irrelevant)
* usr/src/psm/stand/boot/sparc/common/boot_plat.c (irrelevant)

r1mikey commented:
Log of qemu booting with both `kbm_debug` and `prom_debug`: https://gist.github.com/r1mikey/b5850af9fd7a844ee32fc228d688e311

richlowe commented:
I get
```
../../../common/kmdb/kctl/kctl_err.c:54:36: error: too many arguments for format [-Werror=format-extra-args]
../../../common/kmdb/kctl/kctl_err.c:56:36: error: too many arguments for format [-Werror=format-extra-args]
```
That's because `BOP_PUTSARG` explicitly passes an argument, even if there's no format for it, and because I encouraged you to fix `PRINTFLIKE`.

Short of variadic macros, I don't think you can do much here.

I think

```diff
diff --git a/usr/src/cmd/mdb/common/kmdb/kctl/kctl_err.c b/usr/src/cmd/mdb/common/kmdb/kctl/kctl_err.c
index 4fdc701f75..fcad4f7eda 100644
--- a/usr/src/cmd/mdb/common/kmdb/kctl/kctl_err.c
+++ b/usr/src/cmd/mdb/common/kmdb/kctl/kctl_err.c
@@ -51,9 +51,9 @@ kctl_vprintf(int code, const char *format, va_list ap)
                char buf[128];

                if (code == CE_WARN)
-                       BOP_PUTSARG(kctl.kctl_boot_ops, "WARNING: ", NULL);
+                       BOP_PUTSARG(kctl.kctl_boot_ops, "WARNING: ");
                else if (code == CE_NOTE)
-                       BOP_PUTSARG(kctl.kctl_boot_ops, "NOTE: ", NULL);
+                       BOP_PUTSARG(kctl.kctl_boot_ops, "NOTE: ");

                (void) vsnprintf(buf, sizeof (buf), format, ap);
                BOP_PUTSARG(kctl.kctl_boot_ops, "%s\n", buf);
diff --git a/usr/src/uts/aarch64/sys/bootconf.h b/usr/src/uts/aarch64/sys/bootconf.h
index 307bc7ebb5..a76613b86c 100644
--- a/usr/src/uts/aarch64/sys/bootconf.h
+++ b/usr/src/uts/aarch64/sys/bootconf.h
@@ -126,7 +126,7 @@ typedef struct bootops {
 #define        BOP_EALLOC(bop, virthint, size, align, flags)\
                ((bop)->bsys_ealloc)(bop, virthint, size, align, flags)

-#define        BOP_PUTSARG(bop, msg, arg)      ((bop)->bsys_printf)(bop, msg, arg)
+#define        BOP_PUTSARG(bop, ...)   ((bop)->bsys_printf)(bop, __VA_ARGS__)

 #if defined(_KERNEL) && !defined(_BOOT)

diff --git a/usr/src/uts/intel/sys/bootconf.h b/usr/src/uts/intel/sys/bootconf.h
index 311fd40e6f..dd9904dd08 100644
--- a/usr/src/uts/intel/sys/bootconf.h
+++ b/usr/src/uts/intel/sys/bootconf.h
@@ -167,7 +167,7 @@ typedef struct bootops {
 #define        BOP_EALLOC(bop, virthint, size, align, flags)\
                ((bop)->bsys_ealloc)(bop, virthint, size, align, flags)

-#define        BOP_PUTSARG(bop, msg, arg)      ((bop)->bsys_printf)(bop, msg, arg)
+#define        BOP_PUTSARG(bop, ...)   ((bop)->bsys_printf)(bop, __VA_ARGS__)

 #if defined(_KERNEL) && !defined(_BOOT)

```

Is the fix, does that seem fine?

r1mikey commented:
> Is the fix, does that seem fine?

This looks good to me - my apologies for not catching this, I do a lot of partial builds due to being on an underpowered build machine at the moment.

I'll update this in the morning, unless you'd prefer to do the fixup in the merge?

richlowe commented:
It's no problem for me to do it.  Thanks for this!
